---
title: "Type is referenced directly or indirectly in the fulfillment callback of its own 'then' method."
category: error
tags:
- promise
- async
---

When implementing a non-native promise type, if the type references itself in
the `then`` method, this error will occur:

```ts
declare class BadPromise {
  then(
    onfulfilled: (value: BadPromise) => any,
    onrejected: (error: any) => any,
  ): BadPromise;
}

async function test() {
  await new BadPromise();
}
```
